File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;

class Solution {
    public long solution(int n, int[] times) {
        Arrays.sort(times);
        long maxTime = (long)times[times.length - 1] * n;
        long answer = maxTime;
        
        long left = 0;
        long right = maxTime;
        long mid = (left + right) / 2;
        while(left<=right){
            long sum = 0;
            for(int i = 0; i<times.length; i++){
                sum += mid / times[i];
            }
            
            if(n <= sum){
                if(mid < answer){
                    answer = mid;
                }
                right = mid - 1;
                mid = (left + right) / 2;
            }else{
                left = mid + 1;
                mid = (left + right) / 2;
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/3/43238. 입국심사/입국심사.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <climits>

using namespace std;

long long solution(int n, vector<int> times) {
    long long answer = 0;
    
    long maxTime = 0;
    for(auto& time : times){
        if(maxTime<time){
            maxTime = time;
        }        
    }
    
    long long low = 1;
    long long high = maxTime*n;
    while(low <= high){
        long long mid = (low+high)/2;
        long long count = 0;
        for(auto& time : times){
            count += mid / time;
        }
        
        if(count >= n){
            answer = mid;
            high = mid-1;
        }else{
            low = mid+1;
        }
    }
    
    return answer;
}